### ISSUE TITLE
[Discussion] Best practice for adding top level metadata

### ISSUE THREAD

**Issue by idunbarh** [CISA-SBOM-Community/SBOM-Generation#17]
Discussion for Aug 13th 2024 tiger team meeting focused on who best to add top level metadata to an SBOM. For example top level license, supplier, copyright, etc ... which would not be known by an existing sbom generation tool.

The discussion revolved around using `jq` to merge a sbom from a generation tool with additional json data.

For example ...

- [bomctl.cdx.json](https://github.com/user-attachments/files/16606148/bomctl.cdx.json) - a syft generated sbom of the bomctl go binary.
- [metadata.cdx.json](https://github.com/user-attachments/files/16606149/metadata.cdx.json) - a json file containing bomctl specific metadata.

... can be merged using `jq` with the following command ...

```
jq --slurp '.[0] * .[1]' bomctl.cdx.json metadata.cdx.json > merged.jq.cdx.json
```

... to produce this sbom with correct metadata ...

- [merged.jq.cdx.json](https://github.com/user-attachments/files/16606160/merged.jq.cdx.json)

Are there better approaches to this problem?

**Comment by vpetersson** [comment]
I like this approach. My only objection is that we agreed to produce both SPDX **and** CycloneDX and we don't want to maintain *two separate* meta data files (one for CycloneDX and one for SPDX). 

Perhaps we can just alternate between the two and just have a conversion process between them.

Thoughts?

**Comment by riteshnoronha** [comment]
@idunbarh in https://github.com/interlynk-io/sbomasm which allows for spec agnostic merging, we have a feature which allows to specific top-level attributes. Although currently the tool requires a minimum of 2 sboms, we could relax that requirement and act as a spec agnostic enricher. 

The top level attributes can be provided via a config file, or via command-line which could be suitable for CI/CD.  If interested let me know i can make the changes. 

```
➜  ~ sbomasm generate
app:
  name: '[REQUIRED]'
  version: '[REQUIRED]'
  description: '[OPTIONAL]'
  author:
  - name: '[OPTIONAL]'
    email: '[OPTIONAL]'
  primary_purpose: '[REQUIRED]'
  purl: '[OPTIONAL]'
  cpe: '[OPTIONAL]'
  license:
    id: '[OPTIONAL]'
  supplier:
    name: '[OPTIONAL]'
    email: '[OPTIONAL]'
  checksum:
  - algorithm: '[OPTIONAL]'
    value: '[OPTIONAL]'
  copyright: '[OPTIONAL]'
output:
  spec: cyclonedx
  file_format: json
assemble:
  include_dependency_graph: true
  include_components: true
  flat_merge: false
  hierarchical_merge: true
  assembly_merge: false
```

**Comment by vpetersson** [comment]
@riteshnoronha I was literally just toying with `sbomasm` and discovered this. It looks very promising, but it doesn't seem to be possible to merge both container and application SBOMs using `sbomasm`:

```bash
$ sbomasm assemble -m -n "Test" -t "application" -v "0.0.1" container-sbom_cyclonedx.json application-sbom_spdx.json
Error: input sboms are not of the same type
```

Maybe it can still be used for the metadata enrichment only part however.

**Comment by riteshnoronha** [comment]
Yes we don't support cross spec merge as yet, it feels for users to have a seamless experience we need this so people stop thinking about specs and just use sboms. cdx2spdx has a mapping which can be used 

https://docs.google.com/spreadsheets/d/1PIiSYLJHlt8djG5OoOYniy_I-J31UMhBKQ62UUBHKVA/edit?gid=862310124#gid=862310124

**Comment by vpetersson** [comment]
Just for context. I realized the above output tried to merge SPDX and CycloneDX. This was not the intent and has been squared out with @riteshnoronha.

**Comment by riteshnoronha** [comment]
As of sbomasm v0.1.5 we have added an edit command, which allows for metadata editing of an existing sbom. We have described all the details here https://github.com/interlynk-io/sbomasm?tab=readme-ov-file#edit and a complete example here  https://github.com/interlynk-io/sbomasm?tab=readme-ov-file#example. 

One of the advantages i see with the tool vs jq, is its spec agonistic and spec version aware. e.g Tools in CDX changed structure formats between 1.4 and 1.5, a raw json would need to provide examples of each.
